crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+                        // might still be extended in the next track.
+                        None
+                    }
+                } else {
+                    // We shouldn't draw here (stroke of None), so we yield the
+                    // current segment, as it was interrupted.
+                    current_segment.take()
+                };
+                offset += size;
+                interrupted_segment
+            } else {
+                // Reached the end of all tracks, so we interrupt and finish
+                // the current segment.
+                current_segment.take()
+            }
+        },
+    )
+}
+
+/// Returns the correct stroke with which to draw a vline right before column
+/// 'x' when going through row 'y', given the stroke of the user-specified line
+/// at this position, if any.
+/// If the vline would go through a colspan, returns None (shouldn't be drawn).
+/// If the one (when at the border) or two (otherwise) cells to the left and
+/// right of the vline have right and left stroke overrides, respectively,
+/// then the cells' stroke overrides are folded together with the vline's
+/// stroke (with priority to the vline's stroke, followed by the right cell's
+/// stroke, and, finally, the left cell's) and returned. If, however, the cells
+/// around the vline at this row do not have any stroke overrides, then the
+/// vline's own stroke is directly returned.
+pub(super) fn vline_stroke_at_row(
+    grid: &CellGrid,
+    x: usize,
+    y: usize,
+    stroke: Option<Arc<Stroke<Abs>>>,
+) -> Option<Arc<Stroke<Abs>>> {
+    if x != 0 && x != grid.cols.len() {
+        // When the vline isn't at the border, we need to check if a colspan would
+        // be present between columns 'x' and 'x-1' at row 'y', and thus overlap
+        // with the line.
+        // To do so, we analyze the cell right after this vline. If it is merged
+        // with a cell before this line (parent_x < x) which is at this row or
+        // above it (parent_y <= y), this means it would overlap with the vline,
+        // so the vline must not be drawn at this row.
+        let first_adjacent_cell = if grid.has_gutter {
+            // Skip the gutters, if x or y represent gutter tracks.
+            // We would then analyze the cell one column after (if at a gutter
+            // column), and/or one row below (if at a gutter row), in order to
+            // check if it would be merged with a cell before the vline.
+            (x + x % 2, y + y % 2)
+        } else {
+            (x, y)
+        };
+        let Axes { x: parent_x, y: parent_y } = grid
+            .parent_cell_position(first_adjacent_cell.0, first_adjacent_cell.1)
+            .unwrap();
+
+        if parent_x < x && parent_y <= y {
+            // There is a colspan cell going through this vline's position,
+            // so don't draw it here.
+            return None;
+        }
+    }
+
+    let left_cell_stroke = x
```

Comment:
Below, I would expect the cell to win, but it does so only partially. Overriding right and bottom isn't possible. This might be a fundamental consequence of the approach we settled on, but at least not one I was aware of so far and it does feel quite unfortunate.

```typ
#table(
  columns: 2,
  stroke: black,
  table.cell(stroke: red)[a], [b],
  [c], [d],
)
```

Reply:
Yeah... some ideas (posted on Discord, but posting here too to be closer to the context):

1. We could keep track of which cells actually overrode their own strokes to give them priority (we'd draw their strokes on top). Something like a `stroke_overridden` field on `Cell`.
2. We could delay resolution of cell strokes until the lines actually drawn (note that we'd still have to resolve earlier due to show rules, but we could resolve again later). That would also allow us to keep track of which cells overrode their own strokes. The main problem is that this means we would resolve one particular cell's stroke four times more than necessary (once for each side). So I'm not sure this is better than the boolean idea.

Reply:
You know, it's actually interesting that the code below gives the following results.

Either way we should do something about it.

```typ
#table(
  columns: 2,
  stroke: black,
  table.cell(stroke: red)[a], [b],
  [c], [d],
)

#table(
  columns: 2,
  table.cell(stroke: red)[a], [b],
  [c], [d],
)
```

![image](https://github.com/typst/typst/assets/9021226/ca752b00-4fc1-4b47-a793-e8da7e022725)

Reply:
Okay, I just realized that, if we assign a priority to each line segment, the priority can't change mid-segment. This means we will have to break down segments into multiple segments depending on their sources.

Maybe that's generally expected though.

Reply:
Okay, it's done! You might want to take a look at the implementation though, so feel free to unresolve this thread (or create new threads) as needed.

![image](https://github.com/typst/typst/assets/9021226/0c337791-f100-46b6-b46e-9eef319cb419)